### PR TITLE
Allow origin api controllers to be run independently in the master

### PR DESCRIPTION
**Pull request by smarterclayton**
First step to HA.

Run `openshift start master api --config=...` to run just the apiserver.  Run `openshift start master controllers --config=...` to run just the controllers.

Known gaps:
* [ ] no health check or metrics available, need to abstract the code for that
* [ ] no test cases yet

### CODE REVIEW

**Comment by smarterclayton** [comment]
@sdodson fyi, @deads2k or @liggitt or @stevekuznetsov review

**Comment by smarterclayton** [comment]
@ironcladlou so you don't have to jump through hoops:

    openshift start --write-config=config1
    openshift start master --config=config1/master/master-config.yaml
    openshift start master controllers --config=config1/master/master-config.yaml

Two controller loops, one apiserver.

**Comment by smarterclayton** [comment]
@timothysc fyi this is closer to what we discussed.  we'll do the leader lease next.  I'm not going to break controllers up yet, but this gives you two independent pieces

**Comment by timothysc** [comment]
@rrati - lets look over in the a.m.

**Comment by timothysc** [comment]
@sdodson @detiber @brenton @rrati - We should chat as this would change deployment.  Ideally you would want a unit file for each mode.

api 
everything else

This would effect pacemaker config as well.

**Comment by stevekuznetsov** [comment]
@smarterclayton run `hack/update-generated-completions.sh`

File: pkg/cmd/server/start/start_allinone.go [hunk 1]
```diff
@@ -20,22 +20,18 @@ import (
...
-	NodeArgs   *NodeArgs
-
-	CreateCerts      bool
-	ConfigDir        util.StringFlag
-	MasterConfigFile string
-	NodeConfigFile   string
-	PrintIP          bool
-	Output           io.Writer
-	DisabledFeatures []string
+	MasterOptions
```
Comment (deads2k) [inline pkg/cmd/server/start/start_allinone.go]: I really dislike anonymous inclusion because of the unexpected interface fulfillment and name hiding.  Do you object to giving this a name?

**Comment by deads2k** [comment]
Vision question.  If we get HA and active/active working correctly, would we ever want to suggest a configuration other than two full masters (complete with API server and controllers) running together?  That would seem simplify our topology story at the expense of scaling out the API server and controllers independently.  Unless we expect a significant performance degradation due to controller interactions, I'd expect that we'd want to take the simpler approach.

If we'd rather scale out a single thing, then it seems to make sense to support a `json` format config (or use some bash transformations) to modify a config in a bash script using `jq` for our extended tests.

File: pkg/cmd/server/start/start_api.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	"os"
+
+	"github.com/golang/glog"
+	"github.com/spf13/cobra"
+
+	kerrors "k8s.io/kubernetes/pkg/api/errors"
+	kcmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+)
+
+const apiLong = `Start the master controllers
```
Comment (deads2k) [inline pkg/cmd/server/start/start_api.go]: bad description

**Comment by timothysc** [comment]
@deads2k Our plan has always been to go active-active on apiserver and still maintain active-passive on other components, but they become "warmer".  active-active on the api will yield the end user perspective that the system is always running.

However, the extra load that is taken on by existing upstream controllers is cost-prohibitive to go active-active (kcm+sched) because they effect performance with little gain unless the system is in flux (scheduling in progress, updating replicas, etc.) Each additional openshift component of would need to be looked at separately.

**Comment by deads2k** [comment]
> @deads2k Our plan has always been to go active-active on apiserver and still maintain active-passive on other components, but they become "warmer". active-active on the api will yield the end user perspective that the system is always running.

The question is more about how you see a customer configuring this.  Would they be responsible to individually starting api servers and controllers?  If not, this change isn't moving the direction we want.  If so, it's fine.  I think that even an active/passive setup is better served by electing the leader automatically and having a single entity to start up.

**Comment by timothysc** [comment]
The complexity would have to be in the installer... right?

File: pkg/cmd/server/start/start_allinone.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/cmd/server/start/start_allinone.go]: No objection

On Aug 13, 2015, at 9:02 AM, David Eads <notifications@github.com> wrote:

In pkg/cmd/server/start/start_allinone.go
<https://github.com/openshift/origin/pull/4132#discussion_r36970911>:

>  	"github.com/openshift/origin/pkg/cmd/server/start/kubernetes"
>  	cmdutil "github.com/openshift/origin/pkg/cmd/util"
>  )
>
>  type AllInOneOptions struct {
> -	MasterArgs *MasterArgs
> -	NodeArgs   *NodeArgs
> -
> -	CreateCerts      bool
> -	ConfigDir        util.StringFlag
> -	MasterConfigFile string
> -	NodeConfigFile   string
> -	PrintIP          bool
> -	Output           io.Writer
> -	DisabledFeatures []string
> +	MasterOptions

I really dislike anonymous inclusion because of the unexpected interface
fulfillment and name hiding. Do you object to giving this a name?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4132/files#r36970911>.

**Comment by smarterclayton** [comment]
Yeah, three paths:

1. Installer, split
2. Running on infra, split
3. Simple setup and dev, combined.

In the long run, a great installer takes this off the user's plate in a way
that has little downside for us. In the short term, trying to be careful.

Biggest advantage of split is that losing the leader lease lets you shoot
yourself in the head with little impact to users, which means we can keep
API proxy connections open longer.

On Aug 13, 2015, at 10:19 AM, Timothy St. Clair <notifications@github.com>
wrote:

The complexity would have to be in the installer... right?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4132#issuecomment-130693025>.

File: pkg/cmd/server/start/start_api.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+				glog.Fatal(err)
+			}
+		},
+	}
+
+	options.MasterArgs = NewDefaultMasterArgs()
+	options.MasterArgs.StartControllers = false
+
+	flags := cmd.Flags()
+	flags.StringVar(&options.ConfigFile, "config", "", "Location of the master configuration file to run from. Required")
```
Comment (deads2k) [inline pkg/cmd/server/start/start_api.go]: Do we want to require a full `MasterConfig`?  Not all of the information is required just to start the API server.

Also, there are likely to be commands related to masters like, "validate-config", and "write-config".  Would we prefer `openshift master start|[config write|validate|migrate|view-with-defaulting]`?

File: pkg/cmd/server/start/start_api.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+This command starts the controllers for the master.  Running
+
+  $ %[1]s start master controllers
+
+will start the controllers that manage the master state, including the scheduler. The controllers
+will run in the foreground until you terminate the process.`
+
+// NewCommandStartMasterAPI starts only the APIserver
+func NewCommandStartMasterAPI(fullName string, out io.Writer) (*cobra.Command, *MasterOptions) {
```
Comment (deads2k) [inline pkg/cmd/server/start/start_api.go]: `name, fullName string`

File: pkg/cmd/server/start/start_api.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+			}
+		},
+	}
+
+	options.MasterArgs = NewDefaultMasterArgs()
+	options.MasterArgs.StartControllers = false
+
+	flags := cmd.Flags()
+	flags.StringVar(&options.ConfigFile, "config", "", "Location of the master configuration file to run from. Required")
+	BindListenArg(options.MasterArgs.ListenArg, flags, "")
```
Comment (deads2k) [inline pkg/cmd/server/start/start_api.go]: Do we care about the "launching without config" aspect of this command?  It doesn't seem like one we'd be likely to just decide to use.

Comment (smarterclayton) [inline pkg/cmd/server/start/start_api.go]: I thought about it and decided no.  We *want* people to use config.

On Aug 13, 2015, at 10:50 AM, David Eads <notifications@github.com> wrote:

In pkg/cmd/server/start/start_api.go
<https://github.com/openshift/origin/pull/4132#discussion_r36982931>:

> +							fmt.Fprintf(c.Out(), "  %s: %s\n", cause.Field, cause.Message)
> +						}
> +						os.Exit(255)
> +					}
> +				}
> +				glog.Fatal(err)
> +			}
> +		},
> +	}
> +
> +	options.MasterArgs = NewDefaultMasterArgs()
> +	options.MasterArgs.StartControllers = false
> +
> +	flags := cmd.Flags()
> +	flags.StringVar(&options.ConfigFile, "config", "", "Location of the master configuration file to run from. Required")
> +	BindListenArg(options.MasterArgs.ListenArg, flags, "")

Do we care about the "launching without config" aspect of this command? It
doesn't seem like one we'd be likely to just decide to use.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4132/files#r36982931>.

File: pkg/cmd/server/start/start_api.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/cmd/server/start/start_api.go]: Down the road, maybe.  For now, I'd like to keep it simple.

On Thu, Aug 13, 2015 at 10:47 AM, David Eads <notifications@github.com>
wrote:

> In pkg/cmd/server/start/start_api.go
> <https://github.com/openshift/origin/pull/4132#discussion_r36982584>:
>
> > +						for _, cause := range details.Causes {
> > +							fmt.Fprintf(c.Out(), "  %s: %s\n", cause.Field, cause.Message)
> > +						}
> > +						os.Exit(255)
> > +					}
> > +				}
> > +				glog.Fatal(err)
> > +			}
> > +		},
> > +	}
> > +
> > +	options.MasterArgs = NewDefaultMasterArgs()
> > +	options.MasterArgs.StartControllers = false
> > +
> > +	flags := cmd.Flags()
> > +	flags.StringVar(&options.ConfigFile, "config", "", "Location of the master configuration file to run from. Required")
>
> Do we want to require a full MasterConfig? Not all of the information is
> required just to start the API server.
>
> Also, there are likely to be commands related to masters like,
> "validate-config", and "write-config". Would we prefer openshift master
> start|[config write|validate|migrate|view-with-defaulting]?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4132/files#r36982584>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

Comment (smarterclayton) [inline pkg/cmd/server/start/start_api.go]: Comments addressed.

On Thu, Aug 13, 2015 at 11:13 AM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> Down the road, maybe.  For now, I'd like to keep it simple.
>
> On Thu, Aug 13, 2015 at 10:47 AM, David Eads <notifications@github.com>
> wrote:
>
>> In pkg/cmd/server/start/start_api.go
>> <https://github.com/openshift/origin/pull/4132#discussion_r36982584>:
>>
>> > +						for _, cause := range details.Causes {
>> > +							fmt.Fprintf(c.Out(), "  %s: %s\n", cause.Field, cause.Message)
>> > +						}
>> > +						os.Exit(255)
>> > +					}
>> > +				}
>> > +				glog.Fatal(err)
>> > +			}
>> > +		},
>> > +	}
>> > +
>> > +	options.MasterArgs = NewDefaultMasterArgs()
>> > +	options.MasterArgs.StartControllers = false
>> > +
>> > +	flags := cmd.Flags()
>> > +	flags.StringVar(&options.ConfigFile, "config", "", "Location of the master configuration file to run from. Required")
>>
>> Do we want to require a full MasterConfig? Not all of the information is
>> required just to start the API server.
>>
>> Also, there are likely to be commands related to masters like,
>> "validate-config", and "write-config". Would we prefer openshift master
>> start|[config write|validate|migrate|view-with-defaulting]?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/openshift/origin/pull/4132/files#r36982584>.
>>
>
>
>
> --
> Clayton Coleman | Lead Engineer, OpenShift
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

File: pkg/cmd/server/start/start_allinone.go [hunk 6]
```diff
@@ -61,12 +57,8 @@ You may also pass --kubeconfig=<path> to connect to an external Kubernetes clust
 
 // NewCommandStartAllInOne provides a CLI handler for 'start' command
 func NewCommandStartAllInOne(fullName string, out io.Writer) (*cobra.Command, *AllInOneOptions) {
```
Comment (deads2k) [inline pkg/cmd/server/start/start_allinone.go]: Renamed `fullName`.  It isn't a full name in the same sense as our other commands and now that it has a semantic meaning (used below in `DefaultsFromName`).

File: pkg/cmd/server/start/start_api.go [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+This command starts the master API.  Running
+
+  $ %[1]s start master %[2]s
+
+will start the server listening for incoming API requests. The server
+will run in the foreground until you terminate the process.`
+
+// NewCommandStartMasterAPI starts only the APIserver
+func NewCommandStartMasterAPI(name, fullName string, out io.Writer) (*cobra.Command, *MasterOptions) {
```
Comment (deads2k) [inline pkg/cmd/server/start/start_api.go]: Rename `fullName` to reflect what it is.  `baseName`?

File: pkg/cmd/server/start/start_api.go (hunk 5 above)
Comment (deads2k) [inline pkg/cmd/server/start/start_api.go]: > I thought about it and decided no.  We *want* people to use config.

In that case, the `ListenArg` should be extraneous.

File: pkg/cmd/server/start/master_args.go [hunk 8]
```diff
@@ -44,6 +44,8 @@ type MasterArgs struct {
 	// or URL). Defaults to same as --master.
 	MasterPublicAddr flagtypes.Addr
 
+	StartAPI         bool
+	StartControllers bool
```
Comment (deads2k) [inline pkg/cmd/server/start/master_args.go]: Suggestion: Common usage of `NewDefaultMasterArgs` followed by setting the `Start*` bits individually to `false` requires knowledge of all the flags.  Having `NewDefaultMasterArgs` return with the `Start*` bits disabled, will keep things like `openshift start master api` operating as intended if we add more switches.

File: pkg/cmd/server/start/start_controllers.go [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+			}
+		},
+	}
+
+	options.MasterArgs = NewDefaultMasterArgs()
+	options.MasterArgs.StartAPI = false
+
+	flags := cmd.Flags()
+	flags.StringVar(&options.ConfigFile, "config", "", "Location of the master configuration file to run from. Required")
+	BindListenArg(options.MasterArgs.ListenArg, flags, "")
```
Comment (deads2k) [inline pkg/cmd/server/start/start_controllers.go]: I don't think this is required when starting from config.

File: pkg/cmd/server/start/start_master.go [hunk 10]
```diff
@@ -265,11 +265,24 @@ func (o MasterOptions) RunMaster() error {
...
+		openshiftConfig, err := origin.BuildMasterConfig(*masterConfig)
+		if err != nil {
+			return err
+		}
+
+		_, kubeConfig, err := buildKubernetesMasterConfig(openshiftConfig)
+		if err != nil {
+			return err
+		}
+		return StartControllers(openshiftConfig, kubeConfig)
```
Comment (deads2k) [inline pkg/cmd/server/start/start_master.go]: I don't like the way this logic is built.  Why not simply build the `masterConfig` with correct values and have `StartMaster(*origin.MasterConfig)` start the correct pieces as it goes?  That seems more maintainable if we're going to use this as the common entry point.

Comment (deads2k) [inline pkg/cmd/server/start/start_master.go]: If we keep the separate `StartControllers`, I think I'm inclined to also make a separate `StartAPIServer` and call them directly from the commands.

File: pkg/cmd/server/start/start_controllers.go (hunk 9 above)
Comment (smarterclayton) [inline pkg/cmd/server/start/start_controllers.go]: It is, because we need an address to listen on for health check and metrics.  It's unused now but that's in the checklist.  It can't be the value from master because you might run controllers on the same host and you can't bind to the same name.

File: pkg/cmd/server/start/start_api.go (hunk 5 above)
Comment (smarterclayton) [inline pkg/cmd/server/start/start_api.go]: See below

File: pkg/cmd/server/start/start_master.go (hunk 10 above)
Comment (smarterclayton) [inline pkg/cmd/server/start/start_master.go]: Maybe in the future.  But now, it's not worth the larger refactor.  I want to wait until we decide what to do with the UI and the oauth server

Comment (smarterclayton) [inline pkg/cmd/server/start/start_master.go]: Addressed naming and defaulting behavior for StartAPI

On Thu, Aug 13, 2015 at 2:55 PM, David Eads <notifications@github.com>
wrote:

> In pkg/cmd/server/start/start_master.go
> <https://github.com/openshift/origin/pull/4132#discussion_r37011600>:
>
> >  	}
> >
> > -	return nil
> > +	if !o.MasterArgs.StartAPI {
> > +		openshiftConfig, err := origin.BuildMasterConfig(*masterConfig)
> > +		if err != nil {
> > +			return err
> > +		}
> > +
> > +		_, kubeConfig, err := buildKubernetesMasterConfig(openshiftConfig)
> > +		if err != nil {
> > +			return err
> > +		}
> > +		return StartControllers(openshiftConfig, kubeConfig)
>
> If we keep the separate StartControllers, I think I'm inclined to also
> make a separate StartAPIServer and call them directly from the commands.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4132/files#r37011600>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by smarterclayton** [comment]
[test]

File: pkg/cmd/server/start/start_master.go (hunk 10 above)
Comment (deads2k) [inline pkg/cmd/server/start/start_master.go]: > Maybe in the future. But now, it's not worth the larger refactor. I want to wait until we decide what to do with the UI and the oauth server

Can you add a comment saying this behavior shouldn't be extended?  I'm having flashbacks to the old start paths that took over a week to unravel.

File: pkg/cmd/server/start/start_api.go [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+			}
+		},
+	}
+
+	options.MasterArgs = NewDefaultMasterArgs()
+	options.MasterArgs.StartAPI = true
+
+	flags := cmd.Flags()
+	flags.StringVar(&options.ConfigFile, "config", "", "Location of the master configuration file to run from. Required")
+	BindListenArg(options.MasterArgs.ListenArg, flags, "")
```
Comment (deads2k) [inline pkg/cmd/server/start/start_api.go]: I think you'll want comments explaining why these are here or someone will pull them out when they notice.

**Comment by deads2k** [comment]
I'm a little concerned about fragility moving forward without refactoring the start path, but I suppose a couple comments explaining why (and issues open to fix) potentially conflicting flag bindings are present, why the launch path runs on reverse logic that doesn't follow the flags directly, and about why additional bool flags need to rippled through unrelated code would keep me from whining too badly.

I'm assuming this is time critical enough that you aren't going to fix those problems?

**Comment by smarterclayton** [comment]
I don't think we've made the code worse.  Having another split will make
the refactor clearer.  I'll add the comments.

What do you mean by reverse logic?

On Fri, Aug 14, 2015 at 8:34 AM, David Eads <notifications@github.com>
wrote:

> I'm a little concerned about fragility moving forward without refactoring
> the start path, but I suppose a couple comments explaining why (and issues
> open to fix) potentially conflicting flag bindings are present, why the
> launch path runs on reverse logic that doesn't follow the flags directly,
> and about why additional bool flags need to rippled through unrelated code
> would keep me from whining too badly.
>
> I'm assuming this is time critical enough that you aren't going to fix
> those problems?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4132#issuecomment-131093632>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by deads2k** [comment]
> What do you mean by reverse logic?

For the two commands you added, I think the logic would read easier with refactor to allow reading the config and only starting the pieces of the master you want directly in the command.  The current implementation has:
 1. a flag declared in the positive direction: https://github.com/openshift/origin/pull/4132/files#diff-36c6d5003e486e2b6479632137743725R48
 1. a test against the reverse (not): https://github.com/openshift/origin/pull/4132/files#diff-12ac9890607752661a0456afc9ab0ffbR270
 1. a second flag set in the opposing direction: https://github.com/openshift/origin/pull/4132/files#diff-12ac9890607752661a0456afc9ab0ffbR271
 1. another notted check: https://github.com/openshift/origin/pull/4132/files#diff-12ac9890607752661a0456afc9ab0ffbR274
 1. that makes a call a method you'd expect to be called by the other flag: https://github.com/openshift/origin/pull/4132/files#diff-12ac9890607752661a0456afc9ab0ffbR284

It's hard to parse what happens and even then, weird stuff happens.   Consider the "both false" case, where we set a flag, then actually call a startControllers method that will no-op.  

I think that flow is worth refactoring, but I do understand internal beauty for time considerations.

File: pkg/cmd/server/start/start_master.go [hunk 12]
```diff
@@ -314,6 +329,14 @@ func (o MasterOptions) CreateCerts() error {
 	return nil
 }
 
+func buildKubernetesMasterConfig(openshiftConfig *origin.MasterConfig) (bool, *kubernetes.MasterConfig, error) {
+	if openshiftConfig.Options.KubernetesMasterConfig == nil {
+		return false, nil, nil
+	}
+	kubeConfig, err := kubernetes.BuildKubernetesMasterConfig(openshiftConfig.Options, openshiftConfig.RequestContextMapper, openshiftConfig.KubeClient())
```
Comment (liggitt) [inline pkg/cmd/server/start/start_master.go]: call this something other than `kubeConfig` or I will lose my mind

File: pkg/cmd/server/start/start_master.go [hunk 13]
```diff
@@ -371,25 +394,20 @@ func StartMaster(openshiftMasterConfig *configapi.MasterConfig) error {
...
 	}
 
-	var kubeConfig *kubernetes.MasterConfig
-	if openshiftMasterConfig.KubernetesMasterConfig != nil {
-		kubeConfig, err = kubernetes.BuildKubernetesMasterConfig(*openshiftMasterConfig, openshiftConfig.RequestContextMapper, openshiftConfig.KubeClient())
-		if err != nil {
-			return err
-		}
-
+	startKubeMaster, kubeConfig, err := buildKubernetesMasterConfig(openshiftConfig)
```
Comment (liggitt) [inline pkg/cmd/server/start/start_master.go]: `kubeMasterConfig`

File: pkg/cmd/server/start/start_api.go (hunk 11 above)
Comment (liggitt) [inline pkg/cmd/server/start/start_api.go]: I also get twitchy about mixing config and flags... definitely want a comment explaining why this exists, and a TODO to improve it

Comment (smarterclayton) [inline pkg/cmd/server/start/start_api.go]: Comments addressed

On Fri, Aug 14, 2015 at 12:39 PM, Jordan Liggitt <notifications@github.com>
wrote:

> In pkg/cmd/server/start/start_api.go
> <https://github.com/openshift/origin/pull/4132#discussion_r37094201>:
>
> > +							fmt.Fprintf(c.Out(), "  %s: %s\n", cause.Field, cause.Message)
> > +						}
> > +						os.Exit(255)
> > +					}
> > +				}
> > +				glog.Fatal(err)
> > +			}
> > +		},
> > +	}
> > +
> > +	options.MasterArgs = NewDefaultMasterArgs()
> > +	options.MasterArgs.StartAPI = true
> > +
> > +	flags := cmd.Flags()
> > +	flags.StringVar(&options.ConfigFile, "config", "", "Location of the master configuration file to run from. Required")
> > +	BindListenArg(options.MasterArgs.ListenArg, flags, "")
>
> I also get twitchy about mixing config and flags... definitely want a
> comment explaining why this exists, and a TODO to improve it
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4132/files#r37094201>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

File: pkg/cmd/server/start/start_api.go [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+
+	options.MasterArgs = NewDefaultMasterArgs()
+	options.MasterArgs.StartAPI = true
+
+	flags := cmd.Flags()
+	// This command only supports reading from config and the listen argument
+	flags.StringVar(&options.ConfigFile, "config", "", "Location of the master configuration file to run from. Required")
+	cmd.MarkFlagFilename("config", "yaml", "yml")
+	// TODO: add health and metrics endpoints on this listen argument
+	BindListenArg(options.MasterArgs.ListenArg, flags, "")
```
Comment (liggitt) [inline pkg/cmd/server/start/start_api.go]: can we at least let the API use the configured bind address, and just let the controllers command do the weird config+arg thing? controllers would *just* expose health/metrics... if you change the API port here, you probably invalidate a lot of other bits of the config

Comment (smarterclayton) [inline pkg/cmd/server/start/start_api.go]: I'm confused what you're asking.

On Fri, Aug 14, 2015 at 1:26 PM, Jordan Liggitt <notifications@github.com>
wrote:

> In pkg/cmd/server/start/start_api.go
> <https://github.com/openshift/origin/pull/4132#discussion_r37098424>:
>
> > +					}
> > +				}
> > +				glog.Fatal(err)
> > +			}
> > +		},
> > +	}
> > +
> > +	options.MasterArgs = NewDefaultMasterArgs()
> > +	options.MasterArgs.StartAPI = true
> > +
> > +	flags := cmd.Flags()
> > +	// This command only supports reading from config and the listen argument
> > +	flags.StringVar(&options.ConfigFile, "config", "", "Location of the master configuration file to run from. Required")
> > +	cmd.MarkFlagFilename("config", "yaml", "yml")
> > +	// TODO: add health and metrics endpoints on this listen argument
> > +	BindListenArg(options.MasterArgs.ListenArg, flags, "")
>
> can we at least let the API use the configured bind address, and just let
> the controllers command do the weird config+arg thing? controllers would
> *just* expose health/metrics... if you change the API port here, you
> probably invalidate a lot of other bits of the config
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4132/files#r37098424>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by smarterclayton** [comment]
[merge] and will take more review comments in the follow up issues I'll spawn for listening for health and testing.

**Comment by smarterclayton** [comment]
[merge]

File: pkg/cmd/server/start/start_api.go (hunk 14 above)
Comment (deads2k) [inline pkg/cmd/server/start/start_api.go]: > I'm confused what you're asking.

The API server should use the listen port specified in the config file.  The controllers have a reasonable need to respecify, but I'm hard pressed to think of a reason why the API server shouldn't simply use the same port it's using for serving the API.

Comment (smarterclayton) [inline pkg/cmd/server/start/start_api.go]: It should, see the updates where I removed it.

On Mon, Aug 17, 2015 at 9:56 AM, David Eads <notifications@github.com>
wrote:

> In pkg/cmd/server/start/start_api.go
> <https://github.com/openshift/origin/pull/4132#discussion_r37189372>:
>
> > +					}
> > +				}
> > +				glog.Fatal(err)
> > +			}
> > +		},
> > +	}
> > +
> > +	options.MasterArgs = NewDefaultMasterArgs()
> > +	options.MasterArgs.StartAPI = true
> > +
> > +	flags := cmd.Flags()
> > +	// This command only supports reading from config and the listen argument
> > +	flags.StringVar(&options.ConfigFile, "config", "", "Location of the master configuration file to run from. Required")
> > +	cmd.MarkFlagFilename("config", "yaml", "yml")
> > +	// TODO: add health and metrics endpoints on this listen argument
> > +	BindListenArg(options.MasterArgs.ListenArg, flags, "")
>
> I'm confused what you're asking.
>
> The API server should use the listen port specified in the config file.
> The controllers have a reasonable need to respecify, but I'm hard pressed
> to think of a reason why the API server shouldn't simply use the same port
> it's using for serving the API.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4132/files#r37189372>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift
